### PR TITLE
added error messages factory

### PR DESCRIPTION
**Pull request by andriikorotkov**
Improve database sources and destinations error messages
When users are setting up a database connector and fail (due to misconfiguration) they are often provided a non-descriptive error message

The idea of this PR is to map common SqlState code to some human readable messages like: "Incorrect password or user name", "Incorrect host or port"
Thus user may have more information what might be wrong with provided configurations

Examples how it will look on UI:
 - mysql fail database

-postgres fail username

- snowflake fail role

- oracle fail host

- GCS fail bucket name

### CODE REVIEW

File: airbyte-config/config-models/src/main/java/io/airbyte/config/exception/ConnectionFileServiceErrorException.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.config.exception;
+
+public class ConnectionFileServiceErrorException extends RuntimeException {
```
Comment (VitaliiMaltsev) [inline airbyte-config/config-models/src/main/java/io/airbyte/config/exception/ConnectionFileServiceErrorException.java:7]: i did not find usages of this exception. Do we need this class?

Comment (andriikorotkov) [reply airbyte-config/config-models/src/main/java/io/airbyte/config/exception/ConnectionFileServiceErrorException.java]: valid remark. removed it

File: airbyte-db/db-lib/src/main/java/io/airbyte/db/jdbc/DefaultJdbcDatabase.java [hunk 2]
```diff
@@ -73,10 +75,17 @@ public <T> Stream<T> unsafeResultSetQuery(final CheckedFunction<Connection, Resu
...
   public DatabaseMetaData getMetaData() throws SQLException {
-    final Connection conn = dataSource.getConnection();
-    final DatabaseMetaData metaData = conn.getMetaData();
-    conn.close();
-    return metaData;
+    try (final Connection connection = dataSource.getConnection()) {
+      final DatabaseMetaData metaData = connection.getMetaData();
+      return metaData;
+    } catch (SQLException e) {
+      if (Objects.isNull(e.getCause())) {
```
Comment (VitaliiMaltsev) [inline airbyte-db/db-lib/src/main/java/io/airbyte/db/jdbc/DefaultJdbcDatabase.java:82]: please add some comments here in which cases or for which connectors we have _cause==null_

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/errors/ErrorMessageFactory.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+import io.airbyte.integrations.base.errors.messages.DefaultErrorMessage;
+import io.airbyte.integrations.base.errors.messages.ErrorMessage;
+import io.airbyte.integrations.base.errors.messages.MongoDbErrorMessage;
+import io.airbyte.integrations.base.errors.utils.ConnectorName;
+import java.util.Map;
+
+public class ErrorMessageFactory {
+
+  private final static Map<ConnectorName, ErrorMessage> MAP = Map.of(
```
Comment (VitaliiMaltsev) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/errors/ErrorMessageFactory.java:18]: Please rename MAP to CONNECTOR_ERROR_MESSAGE_MAP

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/errors/messages/ErrorMessage.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import io.airbyte.integrations.base.errors.utils.ConnectionErrorType;
+import io.airbyte.integrations.base.errors.utils.ConnectorName;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+public abstract class ErrorMessage {
+
+  private static final String DEFAULT_ERROR_MESSAGE = "State code: %s. \n Error Message: %s.";
+  protected final Map<String, ConnectionErrorType> CONSTANTS = new HashMap<>();
```
Comment (VitaliiMaltsev) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/errors/messages/ErrorMessage.java:16]: the same here. Please rename CONSTANTS to ERRORCODES_TYPES

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/Source.java [hunk 5]
```diff
@@ -36,4 +37,14 @@ public interface Source extends Integration {
...
   AutoCloseableIterator<AirbyteMessage> read(JsonNode config, ConfiguredAirbyteCatalog catalog, JsonNode state) throws Exception;
 
+  /***
+   * This method returns connector name which user for error messages mapping
+   *
+   * @return ConnectorName for example MYSQL or POSTGRES, will return DEFAULT if error message mapping
+   *         is not specified
+   */
+  default ConnectorName getConnectorName() {
+    return ConnectorName.DEFAULT;
```
Comment (VitaliiMaltsev) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/Source.java:47]: i'm not sure that it's right way to return DEFAULT instead of true connectors name. I would prefer return  MYSQL or POSTGRES, etc

File: airbyte-integrations/connectors/destination-gcs/src/main/java/io/airbyte/integrations/destination/gcs/GcsDestination.java [hunk 6]
```diff
@@ -52,10 +56,21 @@ public AirbyteConnectionStatus check(final JsonNode config) {
       S3Destination.testMultipartUpload(s3Client, destinationConfig.getBucketName(), destinationConfig.getBucketPath());
 
       return new AirbyteConnectionStatus().withStatus(Status.SUCCEEDED);
+    } catch (final AmazonS3Exception e) {
+      LOGGER.error("Exception attempting to access the Gcs bucket: {}", e.getMessage());
+      LOGGER.error("Please make sure you account has all of these roles: " + EXPECTED_ROLES);
```
Comment (VitaliiMaltsev) [inline airbyte-integrations/connectors/destination-gcs/src/main/java/io/airbyte/integrations/destination/gcs/GcsDestination.java:61]: I don't think this log is needed here. Lack of permissions is a special case as one of the possible reasons for AmazonS3Exception being thrown, but we need to provide the exact reason

**Review by VitaliiMaltsev — APPROVED** [review]

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/errors/messages/ErrorMessage.java (hunk 4 above)
Comment (andriikorotkov) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/errors/messages/ErrorMessage.java:16]: renamed it

File: airbyte-db/db-lib/src/main/java/io/airbyte/db/jdbc/DefaultJdbcDatabase.java (hunk 2 above)
Comment (andriikorotkov) [reply airbyte-db/db-lib/src/main/java/io/airbyte/db/jdbc/DefaultJdbcDatabase.java:82]: updated it

File: airbyte-integrations/connectors/destination-gcs/src/main/java/io/airbyte/integrations/destination/gcs/GcsDestination.java (hunk 6 above)
Comment (andriikorotkov) [reply airbyte-integrations/connectors/destination-gcs/src/main/java/io/airbyte/integrations/destination/gcs/GcsDestination.java:61]: removed it

**Comment by andriikorotkov** [comment]
@grishick @ryankfu I have seen all your comments, fixed them and updated the error message handling in this pull request. Now the error message that the user will see will consist of the state code, the error code, if any, and the message that comes when the exception is processed.

**Review by grishick — COMMENTED** [review]
Why does MongoDB need a separate MongoDBErrorMessage class that maps some Mongo error codes to error types, but other DBs don't need that?

File: airbyte-integrations/connectors/source-snowflake/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/SnowflakeJdbcSourceAcceptanceTest.java [hunk 7]
```diff
@@ -96,9 +98,34 @@ public AbstractJdbcSource<JDBCType> getJdbcSource() {
...
+
+  @Test
+  public void testCheckIncorrectHostFailure() throws Exception {
+    ((ObjectNode) config).put(JdbcUtils.HOST_KEY, "localhost2");
+    final AirbyteConnectionStatus actual = source.check(config);
+    assertEquals(Status.FAILED, actual.getStatus());
+    assertTrue(actual.getMessage().contains("State code: 28000; Error code: 200028;"));
   }
 
   @Override
```
Comment (ryankfu) [inline airbyte-integrations/connectors/source-snowflake/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/SnowflakeJdbcSourceAcceptanceTest.java:131]: As a side comment, @cgardens has a [draft PR](https://github.com/airbytehq/airbyte/pull/15866/files) that covers some tests not covered here, namely invalid schema and database. Let's try to keep parity with tests since it appears that Postgres and other connectors already cover some of these tests

Also, can you link where the error codes were found? It help to be able to verify where `200028`, `200011`, etc are mentioned in their own documentation. For instance, `390100` is not found within their documentation for [Federated Authentication & SSO Error Codes](https://docs.snowflake.com/en/user-guide/errors-saml.html) so curious how you found these codes?

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/errors/messages/MongoDbErrorMessage.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+import static io.airbyte.integrations.base.errors.utils.ConnectionErrorType.INCORRECT_ACCESS_PERMISSION;
+import static io.airbyte.integrations.base.errors.utils.ConnectionErrorType.INCORRECT_CLUSTER;
+import static io.airbyte.integrations.base.errors.utils.ConnectionErrorType.INCORRECT_HOST_OR_PORT;
+import static io.airbyte.integrations.base.errors.utils.ConnectionErrorType.INCORRECT_USERNAME_OR_PASSWORD_OR_DATABASE;
+import static io.airbyte.integrations.base.errors.utils.ConnectorName.MONGO;
+
+import io.airbyte.integrations.base.errors.utils.ConnectorName;
+
+public class MongoDbErrorMessage extends ErrorMessage {
```
Comment (grishick) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/errors/messages/MongoDbErrorMessage.java:15]: Please add a comment linking to MongoDB doc page that explains these error codes

File: airbyte-integrations/connectors/source-snowflake/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/SnowflakeJdbcSourceAcceptanceTest.java (hunk 7 above)
Comment (andriikorotkov) [reply airbyte-integrations/connectors/source-snowflake/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/SnowflakeJdbcSourceAcceptanceTest.java:131]: @ryankfu as for the tests - I don't think our tests will contradict each other. In some cases, they may cover the same case on different datasets. Therefore, I do not think that it can interfere.

Comment (andriikorotkov) [reply airbyte-integrations/connectors/source-snowflake/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/SnowflakeJdbcSourceAcceptanceTest.java:131]: @ryankfu as for the codes - now I don't process them, but only take the state code either from the **exception** or from **cause of this exception**. This code that I output is returned to me in a `SQLException`.

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/errors/messages/MongoDbErrorMessage.java (hunk 8 above)
Comment (andriikorotkov) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/errors/messages/MongoDbErrorMessage.java]: @grishick I reworked exception handling for mongodb. The error code and error message will now be passed directly from mongo exceptions. As for the `-3 and -4` codes - I didn't find them in the documentation, but I see that these numbers are assigned as an error code in the MongoException class in Mongo driver. I am attaching a screenshot where it occurs.

**Comment by grishick** [comment]
Looks like a couple of destinations are failing tests. If that's the case, a workaround could be to exclude these failing connectors from this PR.

File: airbyte-integrations/connectors/destination-gcs/src/main/java/io/airbyte/integrations/destination/gcs/GcsDestination.java [hunk 9]
```diff
@@ -52,10 +56,18 @@ public AirbyteConnectionStatus check(final JsonNode config) {
       S3Destination.testMultipartUpload(s3Client, destinationConfig.getBucketName(), destinationConfig.getBucketPath());
 
       return new AirbyteConnectionStatus().withStatus(Status.SUCCEEDED);
+    } catch (final AmazonS3Exception e) {
+      LOGGER.error("Exception attempting to access the Gcs bucket: {}", e.getMessage());
```
Comment (bleonard) [inline airbyte-integrations/connectors/destination-gcs/src/main/java/io/airbyte/integrations/destination/gcs/GcsDestination.java:60]: Maybe `AWS` not `Gcs`
Also, please look for other capitalization errors.

File: airbyte-db/db-lib/src/main/java/io/airbyte/db/exception/ConnectionErrorException.java [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+  public String getStateCode() {
+    return this.stateCode;
+  }
+
+  public int getErrorCode() {
+    return errorCode;
+  }
+
+  public String getExceptionMessage() {
+    return exceptionMessage;
```
Comment (ryankfu) [inline airbyte-db/db-lib/src/main/java/io/airbyte/db/exception/ConnectionErrorException.java:46]: How come this return value is not `this.exceptionMessage;`? Also since these values are not set they will default to `null`, no? Is that intended behavior?

File: airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/AbstractSourceConnectorTest.java [hunk 11]
```diff
@@ -133,6 +133,10 @@ protected ConnectorSpecification runSpec() throws WorkerException {
   }
 
   protected StandardCheckConnectionOutput runCheck() throws Exception {
+    return runCheck(getConfig());
+  }
+
+  protected StandardCheckConnectionOutput runCheck(JsonNode config) throws Exception {
```
Comment (ryankfu) [inline airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/AbstractSourceConnectorTest.java:139]: How is `JsonNode config` used here?

Comment (andriikorotkov) [reply airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/AbstractSourceConnectorTest.java:139]: Sorry, but these are not my changes, but github for some reason shows that they are mine

.

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/errors/messages/ErrorMessage.java [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+      return getDefaultErrorMessage(stateCode, exception);
+    } else {
+      return configMessage(stateCode, errorCode, message);
+    }
+  }
+
+  private static String configMessage(String stateCode, int errorCode, String message) {
+    var stateCodePart = Objects.isNull(stateCode) ? "" : "State code: " + stateCode + "; ";
+    var errorCodePart = errorCode == 0 ? "" : "Error code: " + errorCode + "; ";
+    return stateCodePart + errorCodePart + "Message: " + message;
```
Comment (ryankfu) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/errors/messages/ErrorMessage.java:26]: Can we potentially use a `String.format()` here? How come the code is using type inference with `var` since it's clear the string will always be of type string , no?

File: airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/AbstractSourceConnectorTest.java (hunk 11 above)
Comment (andriikorotkov) [reply airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/AbstractSourceConnectorTest.java]: Well noted. Now I no longer use this logic. Deleted it.

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/errors/messages/ErrorMessage.java (hunk 12 above)
Comment (andriikorotkov) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/errors/messages/ErrorMessage.java]: Changed the generation of `errorMessage` using `String.format()`. And using `var` is my habit. Replaced it with String

File: airbyte-integrations/connectors/destination-gcs/src/main/java/io/airbyte/integrations/destination/gcs/GcsDestination.java (hunk 9 above)
Comment (andriikorotkov) [reply airbyte-integrations/connectors/destination-gcs/src/main/java/io/airbyte/integrations/destination/gcs/GcsDestination.java]: Thank you for your comments. Changed it.

File: airbyte-db/db-lib/src/main/java/io/airbyte/db/exception/ConnectionErrorException.java (hunk 10 above)
Comment (andriikorotkov) [reply airbyte-db/db-lib/src/main/java/io/airbyte/db/exception/ConnectionErrorException.java:48]: This is normal method behavior. It just has to return me the message that was set to it. But to prevent NPE, I added the exceptionMessage value by initialization everywhere in the constructors

File: airbyte-integrations/connectors/source-snowflake/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/SnowflakeJdbcSourceAcceptanceTest.java (hunk 7 above)
Comment (ryankfu) [reply airbyte-integrations/connectors/source-snowflake/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/SnowflakeJdbcSourceAcceptanceTest.java:131]: It seems the messaging was lost, the comment was not intended to say that Charles has a PR that would potentially conflict with the changes here. The message was that here are the list of tests that should be covered in each of the different connectors. So namely invalid:
* host
* port
* username
* password
* database
* schema

The reason for asking for how the error/state codes were found is because some of the mongodb tests are asserting something like 
```
assertTrue(airbyteConnectionStatus.getMessage().contains("State code: -4"));
```
However from the standpoint of debugging, if someone wanted to look up what mongo db state code is how can it be determined that they're the right state codes? For instance looking up state codes for mongodb I could only find [these state code](https://www.mongodb.com/docs/v6.0/reference/exit-codes/) but none of them match `-4`. It begs the question, how would these state codes be useful if documentation from mongodb doesn't match the codes in the test and what they mean?

File: airbyte-integrations/connectors/destination-mysql-strict-encrypt/src/test-integration/java/io/airbyte/integrations/destination/mysql/MySqlTestDataComparator.java [hunk 13]
```diff
@@ -30,7 +33,35 @@ protected boolean compareBooleanValues(String firstBooleanValue, String secondBo
...
       return super.compareBooleanValues(firstBooleanValue, secondBooleanValue);
     } else {
-      return super.compareBooleanValues(firstBooleanValue, String.valueOf(secondBooleanValue.equals("1")));
+      return super.compareBooleanValues(firstBooleanValue,
+              String.valueOf(secondBooleanValue.equals("1")));
+    }
+  }
+
+  @Override
+  protected boolean compareDateTimeValues(String expectedValue, String actualValue) {
```
Comment (ryankfu) [inline airbyte-integrations/connectors/destination-mysql-strict-encrypt/src/test-integration/java/io/airbyte/integrations/destination/mysql/MySqlTestDataComparator.java:42]: How is this work relevant to the error messages work?

File: airbyte-integrations/connectors/destination-postgres/src/test/java/io/airbyte/integrations/destination/postgres/PostgresDestinationTest.java [hunk 14]
```diff
@@ -132,6 +140,72 @@ void testDefaultParamsWithSSL() {
...
+        .execute(String.format("create database %s;", DATABASE)));
+    // deny access for database for all users from group public
+    database.execute(connection -> connection.createStatement()
+        .execute(String.format("revoke all on database %s from public;", DATABASE)));
+
+    ((ObjectNode) config).put(JdbcUtils.USERNAME_KEY, USERNAME);
+    ((ObjectNode) config).put(JdbcUtils.PASSWORD_KEY, PASSWORD);
+    ((ObjectNode) config).put(JdbcUtils.DATABASE_KEY, DATABASE);
+
+    var destination = new PostgresDestination();
```
Comment (ryankfu) [inline airbyte-integrations/connectors/destination-postgres/src/test/java/io/airbyte/integrations/destination/postgres/PostgresDestinationTest.java:204]: Can we keep consistency with naming similar to `PostgresDestinationTest#sanityTest`? The test there uses 
```java
final Destination destination = new PostgresDestionation()
```

**Comment by andriikorotkov** [comment]
@grishick the tests for some connectors fail not because of my changes, but because for most of them the changes were mergrd without running the tests. For example, for redshift-destination, no changes were added to secrets in Google Secret Manager. For mysql-destination-strict-encrypt - no changes were added to the transformation of types (when added to mysql-destination they were not added here) and several more connectors with small errors.

File: airbyte-integrations/connectors/source-snowflake/src/main/java/io.airbyte.integrations.source.snowflake/SnowflakeSource.java [hunk 15]
```diff
@@ -113,4 +107,10 @@ private JsonNode buildUsernamePasswordConfig(final JsonNode config, final String
     return Jsons.jsonNode(configBuilder.build());
   }
 
+  @Override
```
Comment (ryankfu) [inline airbyte-integrations/connectors/source-snowflake/src/main/java/io.airbyte.integrations.source.snowflake/SnowflakeSource.java:110]: How come this method was moved since it seems like nothing was changed? Correct me if I'm wrong

**Review by ryankfu — CHANGES_REQUESTED** [review]
There's a few things that still need to be reviewed, the things that seem important are mainly on:
* removing the use of `var` especially in areas where the code explicitly defines a type
* provide a link to each of the connectors state code since it wouldn't be super useful to report a state code for a connector only for ourselves to have difficult finding these codes
* remove the `MySqlTestDataComparator` work since that does not seem relevant to the error messages work

If there's a viewpoint that wasn't clear, let's work together to get this over the finish line. There's a lot a great work here and would love for this to benefit our users

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/errors/messages/MongoDbErrorMessage.java (hunk 8 above)
Comment (ryankfu) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/errors/messages/MongoDbErrorMessage.java]: This is helpful to know though. I'm curious to ask, how useful would exposing `state code` if it seems like the `state codes` are arbitrarily assigned with no real perceptible value to surface them?

**Comment by andriikorotkov** [comment]
> * remove the `MySqlTestDataComparator` work since that does not seem relevant to the error messages work

@ryankfu because of this part, i can't publish my changes to handle messages. This part breaks the tests and makes the normal publishing process impossible. (I try to adhere to the concept, and first run the tests there to publish without turning off the tests)

File: airbyte-db/db-lib/src/main/java/io/airbyte/db/mongodb/MongoDatabase.java [hunk 16]
```diff
@@ -47,6 +49,9 @@ public MongoDatabase(final String connectionString, final String databaseName) {
       this.connectionString = new ConnectionString(connectionString);
       mongoClient = MongoClients.create(this.connectionString);
       database = mongoClient.getDatabase(databaseName);
+    } catch (final MongoConfigurationException e) {
+      LOGGER.error(e.getMessage());
+      throw new ConnectionErrorException(String.valueOf(e.getCode()), e);
```
[the remark this answers is not in the document]
Comment (andriikorotkov) [reply airbyte-db/db-lib/src/main/java/io/airbyte/db/mongodb/MongoDatabase.java:54]: Well noted. Added the message from this exception to the constructor

File: airbyte-db/db-lib/src/main/java/io/airbyte/db/jdbc/DefaultJdbcDatabase.java [hunk 17]
```diff
@@ -76,6 +78,15 @@ public DatabaseMetaData getMetaData() throws SQLException {
...
       final DatabaseMetaData metaData = connection.getMetaData();
       return metaData;
+    } catch (SQLException e) {
+      // For almost all connectors cause != null but, for some connectors, for example redshift cause == null and
+      // in this case we will take the necessary parameters from the exception itself
+      if (Objects.isNull(e.getCause())) {
+        throw new ConnectionErrorException(e.getSQLState(), e.getErrorCode(), e.getLocalizedMessage(), e);
+      } else {
+        var cause = (SQLException) e.getCause();
+        throw new ConnectionErrorException(e.getSQLState(), cause.getErrorCode(), cause.getLocalizedMessage(), cause);
```
[the remark this answers is not in the document]
Comment (andriikorotkov) [reply airbyte-db/db-lib/src/main/java/io/airbyte/db/jdbc/DefaultJdbcDatabase.java:88]: because I call cause multiple times on the next line. I think yes, it looks better. Also I changed `var` to `SQLException` in this case.

File: airbyte-integrations/connectors/destination-gcs/src/main/java/io/airbyte/integrations/destination/gcs/GcsDestination.java [hunk 18]
```diff
@@ -52,10 +56,18 @@ public AirbyteConnectionStatus check(final JsonNode config) {
       S3Destination.testMultipartUpload(s3Client, destinationConfig.getBucketName(), destinationConfig.getBucketPath());
 
       return new AirbyteConnectionStatus().withStatus(Status.SUCCEEDED);
+    } catch (final AmazonS3Exception e) {
+      LOGGER.error("Exception attempting to access the AWS bucket: {}", e.getMessage());
+      var messages = getDefaultErrorMessage(e.getErrorCode(), e);
```
[the remark this answers is not in the document]
Comment (andriikorotkov) [reply airbyte-integrations/connectors/destination-gcs/src/main/java/io/airbyte/integrations/destination/gcs/GcsDestination.java:61]: Yes you're right. I changed the `var` here to `String`, made the `getDefaultErrorMessage` method `private` and will use the `getErrorMessage` method everywhere.

File: airbyte-integrations/connectors/destination-mongodb/src/main/java/io/airbyte/integrations/destination/mongodb/MongodbDestination.java [hunk 19]
```diff
@@ -71,18 +78,35 @@ public AirbyteConnectionStatus check(final JsonNode config) {
...
       final var database = getDatabase(config);
       final var databaseName = config.get(JdbcUtils.DATABASE_KEY).asText();
-      final Set<String> databaseNames = MoreIterators.toSet(database.getDatabaseNames().iterator());
+      final Set<String> databaseNames = getDatabaseNames(database);
       if (!databaseNames.contains(databaseName) && !databaseName.equals(database.getName())) {
         throw new MongodbDatabaseException(databaseName);
       }
       return new AirbyteConnectionStatus().withStatus(AirbyteConnectionStatus.Status.SUCCEEDED);
+    } catch (final ConnectionErrorException e) {
+      var messages = getErrorMessage(e.getStateCode(), e.getErrorCode(), e.getExceptionMessage(), e);
```
[the remark this answers is not in the document]
Comment (andriikorotkov) [reply airbyte-integrations/connectors/destination-mongodb/src/main/java/io/airbyte/integrations/destination/mongodb/MongodbDestination.java:87]: Yes, that's right, it's possible. But in processing, in the `ErrorMessage.getErrorMessage` method, I'll process it, and if the error code is empty, it won't change the readability of the processed error message.

File: airbyte-integrations/connectors/source-snowflake/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/SnowflakeJdbcSourceAcceptanceTest.java (hunk 7 above)
Comment (andriikorotkov) [reply airbyte-integrations/connectors/source-snowflake/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/SnowflakeJdbcSourceAcceptanceTest.java:131]: @ryankfu yes, I agree with you that in some cases posting the code may not give results, but in most cases it is useful information that really helps the user. I can't tell you exactly why the driver developers do this (this is about mongodb), or why not all state codes are in the documentation (this is about your snowflake comment), but for most connectors all state codes are in the documentation and up a normal description is attached to them. This code is intended for general processing - and it handles messages from different connectors in the same way. Unfortunately, I cannot guarantee the accuracy of the state code, or the accuracy of the exception message.

File: airbyte-integrations/connectors/destination-mysql-strict-encrypt/src/test-integration/java/io/airbyte/integrations/destination/mysql/MySqlTestDataComparator.java (hunk 13 above)
Comment (andriikorotkov) [reply airbyte-integrations/connectors/destination-mysql-strict-encrypt/src/test-integration/java/io/airbyte/integrations/destination/mysql/MySqlTestDataComparator.java:42]: this logic is added here in the strict-encrypt connector from the main connector. Here is the number of the pull request and its name in which it was done - `[15245] Destination-mysql: fixed normalization tests after changes in python part`. The developer who added these changes only added them to the main connector, but forgot about strict-encrypt. Therefore, the tests will fail here, which interferes with testing and publishing.

File: airbyte-integrations/connectors/destination-postgres/src/test/java/io/airbyte/integrations/destination/postgres/PostgresDestinationTest.java (hunk 14 above)
Comment (andriikorotkov) [reply airbyte-integrations/connectors/destination-postgres/src/test/java/io/airbyte/integrations/destination/postgres/PostgresDestinationTest.java:204]: updated it

File: airbyte-integrations/connectors/source-snowflake/src/main/java/io.airbyte.integrations.source.snowflake/SnowflakeSource.java (hunk 15 above)
Comment (andriikorotkov) [reply airbyte-integrations/connectors/source-snowflake/src/main/java/io.airbyte.integrations.source.snowflake/SnowflakeSource.java:110]: Sorry, but these are not my changes. I don't know why GitHub displays them

.

**Comment by andriikorotkov** [comment]
@ryankfu I merged changes related to `MySqlTestDataComparator` in a separate pull request. Now they shouldn't get here

File: airbyte-integrations/connectors/destination-mongodb/src/main/java/io/airbyte/integrations/destination/mongodb/MongodbDestination.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (ryankfu) [reply airbyte-integrations/connectors/destination-mongodb/src/main/java/io/airbyte/integrations/destination/mongodb/MongodbDestination.java:86]: Wouldn't that mean that you can be passing in an `uninitialized` int into `getErrorMessage`? That would cause an error cause in `MongoDbSource` you create a `new ConnectionErrorException(String exceptionMessage)` but that does not include an `errorCode`

When I tested this quick snipped out in my IDE this was giving my an error compiling which mimics how `getErrorMessage` works, correct me if I'm wrong
```java
  @Test
  void testUninitialzedIntPassed() {
    final int number;
    System.out.println(runTest("hi", number, "world"));
  }

  public String runTest(final String first, final int num, final String second) {
    final String temp = num == 0 ? "" : " TEST";
    return first + temp + second;
  };
```

**Comment by ryankfu** [comment]
Work was completed in this ticket: https://github.com/airbytehq/airbyte/pull/16202
